### ISSUE TITLE
PERL5LIB conflicts with system perl produce a bad story, what's the solution?

### ISSUE THREAD

**Issue by johnhawkinson** [Homebrew/homebrew-core#10133]
- [x] Ran `brew update` and retried your prior step?
- [x] Ran `brew doctor`, fixed as many issues as possible and retried your prior step?
- [ ] Confirmed this is problem with specific formulae and not Homebrew/brew? If it's a general Homebrew/brew problem please file this issue at https://github.com/Homebrew/brew/issues/new

### Bug report summary
Having installed homebrew's perl package, and followed the guidelines from the formula's Caveats section, I have PERL5LIB set to ~/perl5/lib/perl5, and have some related environment variables similarly set as well.

The result of this is that sometimes the system perl (/usr/bin/perl) breaks mysteriously. Ironically this is often via Homebrew. For instance, upgrading `openssl` failed like this:
```
==> make test
Last 15 lines from /Users/jhawk/Library/Logs/Homebrew/openssl/05.make:
Server's key = 8A7DEB34EAF1EDC085FD2CB54B8F2E8DC9A23D2FCF7B0AD27A1235FE61E6EAFC2
C25ED98B191A98665BE630539F561B69D0BE0462C6A140047B1C962469F21E77BCDC180194885355
61EF6D0AC391CDF70BE4BB787C3E4F883740D094A2A57620673FCE676FEA50AD734BB8186F8DF27B
651E62161F7CE5D71ED77A7B5BA8CFA
CMS consistency test
/usr/bin/perl cms-test.pl
Usage: DynaLoader::dl_find_symbol(libhandle, symbolname) at /Users/jhawk/perl5/l
ib/perl5/darwin-thread-multi-2level/XSLoader.pm line 89.
Compilation failed in require at /System/Library/Perl/5.18/darwin-thread-multi-2
level/IO/Handle.pm line 269.
```

A workaround is to unset PERL5LIB, but that means I don't have access my locally installed perl modules, which is a problem.

How is this supposed to work? Having a $PERL5LIB that breaks /usr/bin/perl doesn't seem a tenable situation.

The system perl is v5.18.2, the homebrew /usr/local/bin/perl is v5.24.0.
Installed perl modules are all in the $PERL5LIB top-level, and $PERL5LIB/5.24.0/ contains only an empty `darwin-thread-multi-2level/` subdir.

### Caveats and .bashrc

Here's what the Caveats section says:
```
pb3:test jhawk$ brew info perl
...
==> Caveats
By default non-brewed cpan modules are installed to the Cellar. If you wish
for your modules to persist across updates we recommend using `local::lib`.

You can set that up like this:
  PERL_MM_OPT="INSTALL_BASE=$HOME/perl5" cpan local::lib
  echo 'eval "$(perl -I$HOME/perl5/lib/perl5 -Mlocal::lib)"' >> ~/.bash_profile
```

And my `.bashrc` actually has:
```
PERL5LIB=~/perl5/lib/perl5 export PERL5LIB
PERL_LOCAL_LIB_ROOT=~/perl5 export PERL_LOCAL_LIB_ROOT
PERL_MB_OPT="--install_base \"/Users/jhawk/perl5\""; export PERL_MB_OPT;
PERL_MM_OPT="INSTALL_BASE=/Users/jhawk/perl5"; export PERL_MM_OPT;
```

### brew config, brew doctor
And here are the requested  `brew config` and `brew doctor` outputs:
```
pb3:~ jhawk$ brew config
HOMEBREW_VERSION: 1.1.10-31-gcd2dd2a
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: cd2dd2a5444bc6ba67f20fcc0d0d113d8005fecb
Last commit: 16 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: fb9ec794a2055f948727b6e16ffc25a83b857537
Core tap last commit: 13 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_REPOSITORY: /usr/local/Homebrew
HOMEBREW_CELLAR: /usr/local/Cellar
HOMEBREW_BOTTLE_DOMAIN: https://homebrew.bintray.com
CPU: quad-core 64-bit broadwell
Homebrew Ruby: 2.0.0-p481
Clang: 7.0 build 700
Git: 2.7.1 => /usr/local/bin/git
Perl: /usr/local/bin/perl => /usr/local/Cellar/perl/5.24.0_1/bin/perl
Python: /usr/bin/python
Ruby: /usr/bin/ruby => /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby
Java: 1.8.0_74, 1.6.0_65-b14-468
macOS: 10.10.5-x86_64
Xcode: 7.2.1
CLT: 7.2.0.0.1.1447826929
X11: 2.7.11 => /opt/X11
pb3:~ jhawk$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry and just ignore them. Thanks!

Warning: You have unlinked kegs in your Cellar
Leaving kegs unlinked can lead to build-trouble and cause brews that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  ghostscript
  lesstif
  numpy
  poppler
  python
pb3:~ jhawk$ 
```

**Comment by ilovezfs** [comment]
If you add
```
ENV.delete("PERL5LIB")
```
to the beginning of the openssl formula's install method does it workaround the problem in that particular case?

**Comment by DomT4** [comment]
It should, I imagine. We already do:
```ruby
# OpenSSL will prefer the PERL environment variable if set over $PATH
# which can cause some odd edge cases & isn't intended. Unset for safety.
ENV.delete("PERL")
```
Almost tempted to say `PERL5LIB` should be unset by default by `superenv` given I went to the trouble of purging `PERL_MB_OPT` and `PERL_MM_OPT` a while back over sandboxing issues.

**Comment by ilovezfs** [comment]
@DomT4 right. I'm wondering if https://github.com/Homebrew/brew/issues/932 will/would fix this.

**Comment by DomT4** [comment]
@ilovezfs Yeah, that seems a safe bet.

**Comment by johnhawkinson** [comment]
Please remember that even if this is correct for everything Homebrew does, it still means that anything that runs /usr/bin/perl explicitly is going to have a problem. A system with a PERL5LIB set that's incompatible with the /usr/bin/perl is asking for trouble. So something is wrong with the setup.

@ilovezfs is certainly right, if the recipe deletes `PERL` then it had better be deleting `PERL5LIB` too. My workaround was to run `PERL5LIB=  brew update openssl`. Should I submit a pull request?

So even if https://github.com/Homebrew/brew/issues/932 is impliemented (and I'd be very hesitant on that...seems like a lot of unintended consquences), there's still a danger by running with the Caveat's recommendation. I'm just not sure what to do about it. I don't want to have to set PERL5LIB by hand for every case where I need it.

**Comment by ilovezfs** [comment]
@johnhawkinson I think the right answer here in the short term is a PR (if you're willing to do it) to Homebrew/brew unsetting PERL5LIB in superenv.

I agree with you that we will need to be cautious with Homebrew/brew#932 but since `stdenv` has basically been deprecated (other than e.g., `scons`) for a while, anything that breaks as a result of that PR would probably be a genuine bug at this point.

**Comment by johnhawkinson** [comment]
PR done.

But I still think something is wrong with the concept of setting PERL5LIB that the perl formula recommends -- it means if I install a package outside of Homebrew (say, from the App Store, or a script I curl|sh download from wherever, or Xcode runs something, etc.) that uses /usr/bin/perl, it can easily break in this way.

I am not sure what the fix is.

Also, while it appears that there are no other formulas that delete $PERL and so trigger this particular problem

```
pb3:Formula jhawk$ pwd
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula
pb3:Formula jhawk$ grep ENV.delete *.rb|grep PERL
openssl.rb:    ENV.delete("PERL")
openssl.rb:    ENV.delete("PERL5LIB")
```

there are several of cases that reference /usr/bin/perl explicitly and can therefore be expected to break in a similar way:

```
pb3:Formula jhawk$ grep /usr/bin/perl *rb
asciiquarium.rb:    inreplace "asciiquarium", "#!/usr/bin/env perl", "#!/usr/bin/perl"
autoconf.rb:    ENV["PERL"] = "/usr/bin/perl"
automake.rb:    ENV["PERL"] = "/usr/bin/perl"
autopsy.rb:+#!/usr/bin/perl -wT
extract_url.rb:    inreplace "extract_url.pl", "#!/usr/bin/env perl", "#!/usr/bin/perl"
innotop.rb:    inreplace "innotop", "#!/usr/bin/env perl", "#!/usr/bin/perl"
man2html.rb:    system "/usr/bin/perl", "install.me", "-batch",
monkeysphere.rb:    inreplace pkgshare/"keytrans", "#!/usr/bin/perl -T",
monkeysphere.rb:                                   "#!/usr/bin/perl -T -I#{libexec}/lib/perl5"
psutils.rb:                         "PERL=/usr/bin/perl",
radare2.rb:      perl_version = `/usr/bin/perl -e 'printf "%vd", $^V;'`
verilator.rb:    system "/usr/bin/perl", bin/"verilator", "-Wall", "--cc", "test.v", "--exe", "test.cpp"
pb3:Formula jhawk$ 
```

**Comment by DomT4** [comment]
Have those actually been tested to see if they break? I run with a `PERL5LIB` set and haven't ran into any problems for a long time, and I've built an awful lot of Homebrew formulae over the last few years 😓.

Note though that this problem isn't exclusive to Homebrew. There are numerous tools for installing different `perl` versions, such as `perlbrew`, that are fairly popular, and it's likely with any custom `perl` you'd want to set at least `PERL5LIB`.

**Comment by johnhawkinson** [comment]
> Have those actually been tested to see if they break

No. That was just intended as a quick sanity check. In a quick test, `asciiquarium` seems to work fine. I'm not sure what a perl script has to do to trigger this kind of incompatibility. Probably to load some common module I have installed in my `PERL5LIB`, so many perl scripts won't trigger the problem, and different people may see different failures depending on what is installed in their `PERL5LIB`. (ugh)

>  it's likely with any custom perl you'd want to set at least `PERL5LIB`.
I almost wonder if /usr/local/bin/perl shouldn't be a wrapper that sets PERL5LIB appropriately. I don't think that's a great idea, but it would avoid some of these issues. I'm not sure how `perlbrew` handles them either...my limited experience has not been great, but that may be unfair of me.

p.s. @ilovezfs:
> I think the right answer here in the short term is a PR (if you're willing to do it) to Homebrew/brew unsetting `PERL5LIB` in superenv.
Sorry, I somehow missed the "in superenv" section of your suggestion when I initially read it. But, as discussed in the PR, I think there's an openssl-specific issue. But sorry about that, I would have written more clearly about it if I had realized I has misread you.

**Comment by ilovezfs** [comment]
@johnhawkinson usually such caveats can be removed and replaced with a wrapper script that sets PERL5LIB and then execs perl. See amtterm asciiquarium clipsafe cloc dpkg extract_url get-flash-videos imapsync innotop moreutils mysql-sandbox mytop oysttyer percona-toolkit percona-xtrabackup pulledpork sql-translator st texapp vit

**Comment by johnhawkinson** [comment]
I'm not sure what caveats @ilovezfs is talking about. I was referring to the idea in the perl recipe's Caveats section that installing stuff with cpan into ~/perl5 was reasonable, along with setting `PERL5LIB` there. I think you meant something else, but I'm not quite sure. And looking at `asciiquarium` I am even more confused, since it appears to set `PERL5LIB` and then invoke the system perl (going so far as to replace `#!/usr/bin/env perl`  with `#!/usr/bin/perl` which as far as I am aware is a recipe for disaster). My assumption is we "just get lucky" there because it doesn't load any of the doom-causing perl modules like `XSLoader`.

---

I just wasted an hour or so trying to track down why `brew install --HEAD ffmpeg` did not install the manpages for ffmpeg. Ultimately I found the answer -- unset PERL5LIB first. @##@@!!!

Maybe I should alias Homebrew (yes, I know, it's a shell function not an alias):
```
function brew() ( echo >&2 Unsetting PERL5LIB to protect you from yourself; PERL5LIB= command brew "$@"; )
```
This is tempting enough that I'm probably going to do it and move on with my life, but it sure feels wrong.

--- 

p.s.: Is there some "trick" to get homebrew to drop me in a build tree with knowledge of how the build was executed? I found myself running `brew install`, copying the ~/Library/Logs/Homebrew/ffmpeg to safe location, uninstalling, running `brew install -i`, munging the logs by hand to escape newlines to get the configure invokation, and messing around in the build tree. And iterating this. The result was each step took a lot longer than I would have liked because ffmpeg takes 6 minutes to build in my machine. (I should be thankful it's not twenty years ago and builds are taking hours...)
